udtube/data/mappers.py

```diff
@@ -143,11 +143,12 @@ def _decode(
         Yields:
             List[str]: decoded symbols.
         """
-        return [
-            vocabulary.get_symbol(c)
-            for c in indices
-            if c not in vocabulary.special_idx
-        ]
+        symbols = []
+        for idx in indices:
+            if idx == special.PAD_IDX:
+                return symbols
```

Comment:
A little confused. If PAD_IDX is 0, and this loop does not go through the iterator backwards, then wouldn't this return early? Isn't the better thing to do in this case to continue and not return?

Reply:
There is a theoretical change in behavior here: it stops generating when it hits pad. 

But in practice, output string lengths are already known to us: they map input string lengths (not counting truncation) so inputs either look like this:

```python
[nonpad_idx, nonpad_idx, ... PAD_IDX, PAD_IDX, PAD_IDX]
```

or in the uncommon case where truncation kicks in:

```python
[nonpad_idx, nonpad_idx, ...., nonpad_idx]
```

In the former case we can halt the second we hit the first padding symbol, safely. Returning early is the point here: it's faster.

The older design (i.e. the one I'm replacing in this PR) is more or less inherited from Yoyodyne, where I really don't know the length of an output string yet during decoding (it's one of the things we're trying to learn) and there we have end-of-string special symbols in addition to padding, and at early stages of training, the system may generate a pad symbol before an end-of-string, so more complex logic is necessary.

I am not sure why it matters what PAD_IDX is (though, yes it's zero).

Reply:
Right, sorry. I forgot how the padding worked in practice, I misconflated it with being n-gram style padding (at the left edge as well). Thanks for the explanation.

Reply:
I see what you mean, yeah. No it's always on the right edge here.